var/spack/repos/builtin/packages/netcdf-cxx4/package.py

```diff
@@ -38,107 +40,40 @@ def flag_handler(self, name, flags):
             flags.append(self.compiler.cc_pic_flag)
         if name == "cxxflags" and "+pic" in self.spec:
             flags.append(self.compiler.cxx_pic_flag)
-        elif name == "ldlibs":
-            # Address the underlinking problem reported in
-            # https://github.com/Unidata/netcdf-cxx4/issues/86, which also
-            # results into a linking error on macOS:
-            flags.append(self.spec["netcdf-c"].libs.link_flags)
-
-        # Note that cflags and cxxflags should be added by the compiler wrapper
-        # and not on the command line to avoid overriding the default
-        # compilation flags set by the configure script:
+
         return flags, None, None
 
     @property
     def libs(self):
         libraries = ["libnetcdf_c++4"]
+        shared = "+shared" in spec
 
-        query_parameters = self.spec.last_query.extra_parameters
-
-        if "shared" in query_parameters:
-            shared = True
-        elif "static" in query_parameters:
-            shared = False
-        else:
-            shared = "+shared" in self.spec
-
-        libs = find_libraries(libraries, root=self.prefix, shared=shared, recursive=True)
-
-        if libs:
-            return libs
+        return find_libraries(libraries, root=self.prefix, shared=shared, recursive=True)
```

Comment:
(minor) As originally implemented, this would have allowed for `spec[netdf-cxx4:shared].libs`, although I don't see any packages making use of that, I also don't think that this is necessary for the other changes. Did you encounter an issue with this while you were making other changes?

Reply:
(major) The original implementation also returns `None` if the libraries are not found. It's worse than throwing [an exception](https://github.com/spack/spack/blob/2d8e0825feb05f082f5dc4b4c3832fc323f7cf07/var/spack/repos/builtin/packages/netcdf-fortran/package.py#L111-L114) but much better than an empty list.

Reply:
@scheibelp I kept this due to the library name being different from the package name, which, afaik, requires a custom libs property.

@skosukhin I removed this as it was not clear to me that throwing an exception here is expected spack behaviour. Specifically, [the](https://spack.readthedocs.io/en/latest/packaging_guide.html#custom-attributes) [docs](https://spack-tutorial.readthedocs.io/en/latest/tutorial_advanced_packaging.html#providing-libraries-to-dependents) don't recommend it, and most spack packages don't do this:

```
$ ag -A60 "libs\(self\)\:" | grep -i "spack.error"
netcdf-fortran/package.py:112-        raise spack.error.NoLibrariesError(
python/package.py:1054-        raise spack.error.NoLibrariesError(
libszip/package.py:35-            raise spack.error.NoLibrariesError(
parallel-netcdf/package.py:101-        raise spack.error.NoLibrariesError(
libaec/package.py:48-            raise spack.error.NoLibrariesError(
eccodes/package.py:301-        raise spack.error.NoLibrariesError(
hdf/package.py:126-            raise spack.error.NoLibrariesError(
serialbox/package.py:133-        raise spack.error.NoLibrariesError(msg.format(self.spec.name, self.spec.prefix))
```
Indeed, the only that do are packages maintained by you. Thus, I was following the broader community practice of not including it. I can see value in the error, and ultimately I really don't care. I can add it back if this is something you feel very strongly about.

Reply:
The broader community practice is to rely on the [default handler](https://github.com/spack/spack/blob/de3d1e6c66896f90f090c298fa05c5aefb42bde8/lib/spack/spack/spec.py#L1109C5-L1109C26), which raises [`NoLibrariesError`](https://github.com/spack/spack/blob/de3d1e6c66896f90f090c298fa05c5aefb42bde8/lib/spack/spack/spec.py#L1160).

Reply:
Fair. I've made the change